### PR TITLE
soundwire: cadence: recheck device0 attachment after status change

### PR DESCRIPTION
**Pull request by plbossart**
This patch adds a status check after device0 attachment to solve race
conditions observed during attachment with multiple devices per link

The sequence is the following

1) deviceA attaches as device0

2) the hardware detects a device0 status change and throws an
 interrupt.

3) the interrupt handler schedules the work function

4) the workqueue starts, we read the status
slave0 = cdns_readl(cdns, CDNS_MCP_SLAVE_INTSTAT0);
slave1 = cdns_readl(cdns, CDNS_MCP_SLAVE_INTSTAT1);

we deal with the status change and program deviceA device number to a
non-zero value.

5) deviceB attaches as device0, the device0 status seen by the
hardware does not change.

6) we clear the CDNS_MCP_SLAVE_INTSTAT0/1 registers -> we will never detect
deviceB!

This patch suggest re-checking in a loop the device0 status with a
PING frame, i.e. using the real device0 status instead of information
on status changes.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

### CODE REVIEW

**Comment by marc-hb** [comment]
Is this related to some existing bug(s) already filed? If yes then the links would be useful IMHO.

**Comment by plbossart** [comment]
> Is this related to some existing bug(s) already filed? If yes then the links would be useful IMHO.

yeah, forgot to add
BugLink: https://github.com/thesofproject/linux/issues/3063

**Comment by RanderWang** [comment]
Why once the device is unattached , it will never be attached again ?

File: drivers/soundwire/cadence_master.c [hunk 1]
```diff
@@ -968,10 +970,45 @@ static void cdns_update_slave_status_work(struct work_struct *work)
...
 	cdns_update_slave_status(cdns, slave_intstat);
 	cdns_writel(cdns, CDNS_MCP_SLAVE_INTSTAT0, slave0);
 	cdns_writel(cdns, CDNS_MCP_SLAVE_INTSTAT1, slave1);
 
+	/*
+	 * When there is more than one peripheral per link, it's
+	 * possible that a deviceB becomes attached after we deal with
+	 * the attachment of deviceA. Since the hardware does a
+	 * logical AND, the attachment of the second device does not
+	 * change the status seen by the driver.
```
Comment (RanderWang) [inline drivers/soundwire/cadence_master.c:983]: is it possible for two devices become unattached like this case ?

**Comment by plbossart** [comment]
> Why once the device is unattached , it will never be attached again ?

thats' the point, it *may* reattach but if it does so before we clear the interrupt status bits we will *never* detect it.
If the device re-attaches *after* we clear the interrupt status, we should see it come back as attached.

File: drivers/soundwire/cadence_master.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/cadence_master.c:983]: it both devices become unattached after we program the device number to a non-zero value (so in theory there are no device0 attached any longer), and reattach before we clear the interrupt status, yes we would not detect either as well.

**Review by RanderWang — APPROVED** [review]
LGTM

**Comment by plbossart** [comment]
ok, let's merge this and see if this changes the behavior on Volteer+TGL RVP.

@bardliao let's wait to upstream this for now, we need at least a couple of weeks of test results.

**Pull request merged**